odoo/http.py

```diff
@@ -281,8 +281,10 @@ def __exit__(self, exc_type, exc_value, traceback):
                     self._cr.commit()
                     if self.registry:
                         self.registry.signal_changes()
-                elif self.registry:
-                    self.registry.reset_changes()
+                else:
+                    if self.registry:
+                        self.registry.reset_changes()
+                    self._cr.rollback()
```

Comment:
For the sake of readability, I would make it more similar to the `commit` case above:
```suggestion
                    self._cr.rollback()
                    if self.registry:
                        self.registry.reset_changes()
```